[export] Initial serialization v2

angelayi opened the pull request:
v2 of https://github.com/pytorch/pytorch/pull/102125 because of git issues
corresponding deserialization diff: https://github.com/pytorch/pytorch/pull/102716

Implementing serialization of the exported program to a python dataclass, and then from that dataclass to json. This is split into a couple of sections:
- `serialize(ep: ep.ExportedProgram, opset_version: Dict[str, int]) -> Tuple[bytes, bytes]` -- takes an exported program object, a dictionary mapping opset namespaces to versions, and returns the serialized exported program in bytes, and separately the state dict serialized in bytes
- `GraphModuleSerializer` class that serializes torch.fx.GraphModule 
to the schema.GraphModule dataclass
- `ExportedProgramSerializer` class that serializes torch._export.exported_program.ExportedProgram to the schema.ExportedProgram dataclass

Serialization TODOs:
- [x] pytree spec: https://github.com/pytorch/pytorch/pull/102577
- [ ] higher order ops 
- [ ] node metadata (specifically nn_module_stack/source_fn) https://github.com/pytorch/pytorch/pull/103274
- [ ] constraints https://github.com/pytorch/pytorch/pull/103273
- [ ] graph module metadata

The tests are not super comprehensive, but that's because I think it'll be better tested + easier to test once deserialization is implemented.

ydwu4 reviewed (commented):
Can add some high-level descriptions about this PR in the pr description? Just to keep a record in case we need to find out why we do it in this way in the future and also encourage people to add more meaningful reviews.Example description could include 1. what's the motivation of this change? 2. how users are supposed to user this API 3. Does the test cover all corner cases? Anything that's in your mind but not tested yet? 4. what are the implementation choices we made and why we choose this implementation instead of the alternatives? 5. Is there any lilmitations of current design? Feel free to skip some if you find meaningless and it would be better if new points could be added if you feel that may be helpful for people to understand better. 

I understand it's discussed offline and adding more descriptions could slow down our development a little bit but would definitely be helpful to keep everyone (e.g. those who didn't participate in the discussion like me) on the same page and help us maintain the code in the long run :).

angelayi committed:
[export] Initial serialization v2 (#102707)

Summary:
v2 of https://github.com/pytorch/pytorch/pull/102125 because of git issues
corresponding deserialization diff: https://github.com/pytorch/pytorch/pull/102716

Implementing serialization of the exported program to a python dataclass, and then from that dataclass to json. This is split into a couple of sections:
- `serialize(ep: ep.ExportedProgram, opset_version: Dict[str, int]) -> Tuple[bytes, bytes]` -- takes an exported program object, a dictionary mapping opset namespaces to versions, and returns the serialized exported program in bytes, and separately the state dict serialized in bytes
- `GraphModuleSerializer` class that serializes torch.fx.GraphModule
to the schema.GraphModule dataclass
- `ExportedProgramSerializer` class that serializes torch._export.exported_program.ExportedProgram to the schema.ExportedProgram dataclass

Serialization TODOs:
- [x] pytree spec: https://github.com/pytorch/pytorch/pull/102577
- [ ] higher order ops
- [ ] node metadata (specifically nn_module_stack/source_fn)
- [ ] constraints
- [ ] graph module metadata

The tests are not super comprehensive, but that's because I think it'll be better tested + easier to test once deserialization is implemented.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/102707

Reviewed By: zhxchen17

Differential Revision: D46362466

Pulled By: angelayi

fbshipit-source-id: 8627d9f783cea5af9c36b09f4216c7effc021593
